### ISSUE TITLE
[BUG] Downloaded playlists in wrong order

### ISSUE THREAD

**Issue by AhegaHOE** [jmshrv/finamp#689]
After updating to version `0.6.20` a new bug occurred where downloaded songs in a playlist appear in the order that they were downloaded in rather than the original order of the playlist.

**Comment by Chaphasilor** [comment]
Is this still a problem with 0.6.23?

**Comment by nathan-contino** [comment]
Still happening in downloaded playlists and, worse, albums in 0.6.23.

**Comment by Chaphasilor** [comment]
@nathan-contino that is expected, and there's no good way to fix the order of albums downloaded with the affected versions. You'll have to re-download those.  

I also recommend you switch to the beta version, that has a brand new download system that is much more capable than the old ones (more reliable, download artists/genres/favorites, better synchronization, downloads metadata, can transcode downloads, faster, etc.). Especially if you'd need to download a lot, just reset the app or uninstall it (to get rid of the broken downloads), and then install the beta.  
Theoretically you can migrate old downloads to the new system, but that won't fix the order either.

**Comment by Komodo5197** [comment]
Actually, in theory the migration to the beta should catch the incorrect order during the downloads repair and update it to match the server.

**Comment by nathan-contino** [comment]
Gotcha, I've been tempted to try out the beta for a while now so it's probably due time that I try it out.

Just checking before I blow everything away and start anew: as an F-Droid user, there's no way for me to simply replace the app with the beta version, right? Instead, I have to delete the prod version and likely restart my downloads collection from scratch.

Sidenote: it might be nice if the beta version and the prod version could exist side by side as separate packages with distinct names so I could try the beta without losing everything. Though I'm sure you have your reasons for not doing that!

**Comment by Chaphasilor** [comment]
Correct. In the end it boils down to application IDs and signatures.

F-Droid builds apps right from code (likely to ensure the public source code is actually what ends up in the final app users download), and for that they use their own signature for signing the build (which verifies that the app was actually built by them). We don't have access to their signature, and \@jmshrv's signature is used for the APKs uploaded to GitHub and the Play Store instead.  
If you want to make sure, you can try downloading the beta version from GitHub and installing it as an update; Android should not let you do it.

There's no beta on F-Droid because we would need to create a completely separate app called "Finamp Beta", and set up all of the build steps for that too (otherwise the highest version number is always considered the latest version, even if it's a beta version). Theoretically it's possible to set up this beta app, but I don't have the time to mess with it. If you want to give it a try, feel free to do that, but since it's a separate app you *still* wouldn't be able to update it (it would install alongside the stable version).

The Play Store supports beta versions, but requires the beta to use the *same* application ID, so that it's possible to update the app. It doesn't support installing a beta version alongside the stable version.

So in summary, you cannot upgrade from the F-Droid version to the beta version. You can either uninstall the stable version and install the app from GitHub (keeping it updated using something like Obtanium). Or I can generate you a custom version of the beta that you can install alongside the stable version. You won't be able to update it, but it would allow you to try it out first.

**Comment by nathan-contino** [comment]
Tried the beta, thanks for the tip! OPUS transcoded downloads are saving me a ton of local space, the new interface looks great, and the brand new download system is indeed better/faster/stronger as advertised. Thanks for pushing me to try the beta, I'm really glad I took the plunge. Once we get Android Auto this will be my favourite music player ever.

You can probably mark this issue as resolved, BTW -- seems like the latest stable version always downloads in the correct order, as does beta, so all folks have to do is delete the wrong-ordered downloads and re-download them to fix the issue. Thanks!

**Comment by Chaphasilor** [comment]
Awesome, I'm very glad to hear that! I'll try to get initial Android Auto support squished into an update soon, it's basically working. I just need to fix some voice commands when in offline mode...
